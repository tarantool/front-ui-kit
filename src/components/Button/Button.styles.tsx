import { css } from '@emotion/css';

import { baseFontFamily, colors } from '../../variables';

export const styles = {
  button: css`
    white-space: nowrap;
    border: none;
    border-radius: 4px;
    box-sizing: border-box;
    font-family: ${baseFontFamily};
    font-size: 14px;
    line-height: 22px;
    transition-timing-function: ease-in-out;
    transition-duration: 0.07s;
    transition-property: background-color, color;
    outline: none;
    cursor: pointer;
    -webkit-font-smoothing: antialiased;

    &:hover,
    &:focus,
    &:active {
      transition-property: background-color, color, box-shadow;
    }

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      cursor: default;
    }
  `,

  icon: css`
    width: 16px;
    height: 16px;
    margin-bottom: 2px;
  `,

  iconMargin: css`
    margin-right: 8px;
  `,

  iconRightMargin: css`
    margin-left: 8px;
  `,

  loading: css`
    position: relative;
    color: rgba(0, 0, 0, 0);
    transition: none;

    &:hover,
    &:focus,
    &:active {
      cursor: default;
      color: rgba(0, 0, 0, 0);
    }

    & > *,
    &:hover > *,
    &:active > *,
    &:focus > * {
      visibility: hidden;
    }

    & > *:last-child {
      visibility: visible;
    }
  `,

  loadingWrap: css`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: flex;
    justify-content: center;
    align-items: center;
  `,

  l: css`
    padding: 9px 20px;
  `,

  m: css`
    padding: 5px 16px;
  `,

  s: css`
    padding: 1px 16px;
  `,

  xs: css`
    padding: 1px 9px;
    line-height: 18px;
    font-size: 12px;
  `,

  iconicL: css`
    padding: 9px 12px;
  `,

  iconicM: css`
    padding: 5px 8px;
  `,

  iconicS: css`
    padding: 1px 4px;
  `,
};

export const intentStyles = {
  base: css`
    background-color: white;
    color: ${colors.dark65};
    box-shadow: inset 0 0 0 1px ${colors.baseBg};

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      color: ${colors.intentBase};
      background-color: ${colors.intentBaseActive};
      box-shadow: inset 0 0 0 1px ${colors.intentBase};
    }

    & svg {
      fill: ${colors.dark65};
    }

    &:disabled svg {
      fill: ${colors.intentBase};
    }
  `,

  dark: css`
    background-color: ${colors.dark40};
    color: #ffffff;

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      color: ${colors.intentBase};
      background-color: ${colors.intentBaseActive};
    }

    & svg {
      fill: #ffffff;
    }

    &:disabled svg {
      fill: ${colors.intentBase};
    }
  `,

  primary: css`
    background-color: ${colors.intentPrimary};
    color: #ffffff;

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      background-color: ${colors.intentPrimaryDisabled};
      color: #ffffff;
    }

    & svg {
      fill: #ffffff;
    }

    &:disabled svg {
      fill: #ffffff;
    }
  `,

  secondary: css`
    background-color: ${colors.dark10};
    color: ${colors.dark65};

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      background-color: ${colors.intentBaseActive};
      color: ${colors.intentBase};
      box-shadow: inset 0 0 0 1px ${colors.intentBase};
    }

    & svg {
      fill: ${colors.dark65};
    }

    &:disabled svg {
      fill: ${colors.intentBase};
    }
  `,

  plain: css`
    background-color: transparent;
    color: ${colors.dark65};

    &:disabled,
    &:disabled:active,
    &:disabled:hover {
      background-color: ${colors.intentBaseActive};
      color: ${colors.intentBase};
    }

    & svg {
      fill: ${colors.dark65};
    }

    &:disabled svg {
      fill: ${colors.intentBase};
    }
  `,
};

export const intentActiveStyles = {
  base: css`
    &:focus,
    &:hover {
      background-color: ${colors.dark10};
      box-shadow: none;
    }

    &:active {
      background-color: ${colors.dark10};
      box-shadow: inset 0 4px 0 rgba(4, 11, 29, 0.1);
    }
  `,

  dark: css`
    &:focus,
    &:hover {
      background-color: ${colors.dark60};
    }

    &:active {
      background-color: ${colors.dark60};
      box-shadow: inset 0 4px 0 rgba(4, 11, 29, 0.1);
    }
  `,

  primary: css`
    &:focus,
    &:hover {
      background: ${colors.intentPrimaryActive};
    }

    &:active {
      background: ${colors.intentPrimaryActive};
      box-shadow: inset 0 4px 0 rgba(4, 11, 29, 0.1);
    }
  `,

  secondary: css`
    &:focus,
    &:hover {
      background-color: ${colors.dark25};
    }

    &:active {
      background-color: ${colors.dark25};
      box-shadow: inset 0 4px 0 rgba(4, 11, 29, 0.1);
    }
  `,

  plain: css`
    &:focus,
    &:hover {
      background-color: ${colors.dark10};
    }
  `,
};
