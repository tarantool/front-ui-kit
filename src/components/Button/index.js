// @flow
import * as React from 'react';
import { css, cx } from 'emotion'

const styles = {
  btn: css`
    border: solid 1px gray;
  `
}

type ButtonProps = {
  className?: string,
  children?: React.Node,
  onClick?: (MouseEvent) => void
};

export default ({ children, className, ...props }: ButtonProps) => (
  <button className={cx(className, styles.btn)} {...props}>{children}</button>
);
