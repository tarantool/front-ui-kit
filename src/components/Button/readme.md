Button example:

```js
<Button>Click me</Button>
```
