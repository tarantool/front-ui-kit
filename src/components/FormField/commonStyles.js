import { css } from '@emotion/css';

export const commonFormFieldStyles = {
  wrap: css`
    margin-bottom: 10px;
  `,
  wrapMargin: css`
    margin-bottom: 20px;
  `,
  tooltip: css`
    display: inline-block;
    margin-left: 8px;
  `,
  headingPane: css`
    display: flex;
    flex-direction: row;
    align-items: center;
    height: 22px;
    margin-bottom: 0;
  `,
  headingPaneMargin: css`
    margin-bottom: 10px;
  `,
  headingPaneLeft: css`
    display: flex;
    flex-direction: row;
    align-items: baseline;
  `,
  subTitle: css`
    margin-left: 32px;
  `,
  label: css`
    display: block;
  `,
  topRightControls: css`
    margin-left: auto;
  `,
};
