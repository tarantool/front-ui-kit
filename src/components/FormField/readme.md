It's not recommended to use `topRightControls` without `largeMargins`.

```js
import { useState } from 'react';
import { css } from '@emotion/css';
import {
  Button,
  InputGroup,
  RadioButton,
  Switcher
} from '@tarantool.io/ui-kit';

const styles = {
  wrap: css`
    display: flex;
    flex-direction: row-reverse;
  `,
  demoPanel: css`
    flex-grow: 1;
  `,
  controls: css`
    flex-basis: 200px;
    align-self: flex-start;
    margin-left: 30px;
  `
};

const [value, setValue] = useState(null);

const [flags, setFlags] = useState({
  info: false,
  largeMargins: false,
  rightControls: false,
  subtitle: false,
});

const handleChange = value => setValue(value);
const toggle = key => () => setFlags({ ...flags, [key]: !flags[key] });

const vshard_groups = [
  { name: 'group 1' },
  { name: 'group 2' },
  { name: 'group 3' },
  { name: 'group 4' }
];

<div className={styles.wrap}>
  <InputGroup className={styles.controls}>
    {[
      <Switcher
        checked={flags.info}
        onChange={toggle('info')}
      >
        Info tooltip
      </Switcher>,
      <Switcher
        checked={flags.largeMargins}
        onChange={toggle('largeMargins')}
      >
        Large margins
      </Switcher>,
      <Switcher
        checked={flags.rightControls}
        onChange={toggle('rightControls')}
      >
        Right controls
      </Switcher>,
      <Switcher
        checked={flags.subtitle}
        onChange={toggle('subtitle')}
      >
        Subtitle
      </Switcher>
    ]}
  </InputGroup>
  <div className={styles.demoPanel}>
    <FormField
      label='VShard group'
      largeMargins={flags.largeMargins}
      info={flags.info && 'Extended text description'}
      subTitle={flags.subtitle && 'Subtitle'}
      topRightControls={flags.rightControls && [
        <Button text='Right' intent='secondary' />,
        <Button text='Contorls' intent='secondary' />,
        <Button text='Pane' intent='secondary' />
      ]}
    >
      {vshard_groups && vshard_groups.map(({ name }) => (
        <RadioButton
          onChange={() => handleChange(name)}
          name='vshard_group'
          value={name}
          checked={name === value}
        >
          {name}
        </RadioButton>
      ))}
    </FormField>

    <FormField
      label='VShard group'
      largeMargins={flags.largeMargins}
      info={flags.info && 'Extended text description'}
      subTitle={flags.subtitle && 'Subtitle'}
      topRightControls={flags.rightControls && [
        <Button text='Right' intent='secondary' />,
        <Button text='Contorls' intent='secondary' />,
        <Button text='Pane' intent='secondary' />
      ]}
    >
      {vshard_groups && vshard_groups.map(({ name }) => (
        <RadioButton
          onChange={() => handleChange(name)}
          name='vshard_group'
          value={name}
          checked={name === value}
        >
          {name}
        </RadioButton>
      ))}
    </FormField>
  </div>
</div>
```
