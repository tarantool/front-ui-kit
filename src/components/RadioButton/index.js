// @flow
import * as React from 'react';
import { css, cx } from '@emotion/css';
import { rgba } from 'emotion-rgba';
import { colors } from '../../variables';
import { IconRadio } from '../Icon';
import { Text } from '../Text';

const styles = {
  icon: css`
    display: block;
  `,
  iconWrap: css`
    position: relative;
  `,
  children: css`
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  `,
  childrenMargin: css`
    margin-right: 8px;
  `,
  input: css`
    position: absolute;
    clip: rect(0 0 0 0);
    width: 1px;
    height: 1px;
    margin: -1px;
    appearance: none;

    & + div::before {
      content: '';
      position: absolute;
      top: -2px;
      left: -2px;
      right: -2px;
      bottom: -2px;
      border: solid 1px rgba(255, 255, 255, 0);
      border-radius: 50%;
    }

    &:focus + div::before {
      border-color: ${rgba(colors.intentPrimary, 0.55)};
    }
  `,
  label: css`
    display: flex;
    align-items: center;
    cursor: pointer;
  `,
};

type RadioButtonProps = {
  checked?: boolean,
  children?: React.Node,
  className?: string,
  disabled?: boolean,
  name?: string,
  title?: string,
  value?: string,
  onChange?: (MouseEvent) => void,
};

export const RadioButton = ({
  checked,
  children,
  className,
  disabled,
  onChange,
  name,
  title,
  value,
}: RadioButtonProps) => (
  <label className={cx(styles.label, className)} title={title}>
    <input
      checked={checked}
      className={styles.input}
      disabled={disabled}
      type="radio"
      onChange={onChange}
      name={name}
      value={value}
    />
    <div className={cx(styles.iconWrap, { [styles.childrenMargin]: children })}>
      <IconRadio className={styles.icon} checked={checked} disabled={disabled} />
    </div>
    {typeof children === 'string' ? <Text className={styles.children}>{children}</Text> : children}
  </label>
);
