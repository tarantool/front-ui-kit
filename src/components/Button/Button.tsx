
import React, { ReactNode, MouseEvent, forwardRef, ComponentType } from 'react';
import { cx } from '@emotion/css';

import { IconSpinner } from '../Icon';
import { styles, intentStyles, intentActiveStyles } from './Button.styles';

export interface ButtonIconProps {
  className?: string;
}

export interface ButtonProps {
  autoFocus?: boolean,
  className?: string,
  children?: ReactNode,
  disabled?: boolean,
  icon?: ComponentType<ButtonIconProps>,
  iconRight?: ComponentType<ButtonIconProps>,
  intent?: 'primary' | 'secondary' | 'base' | 'plain' | 'dark',
  onClick?: (e: MouseEvent<HTMLButtonElement>) => void,
  loading?: boolean,
  size?: 'l' | 's' | 'xs' | 'm',
  text?: string,
  title?: string,
  type?: 'button' | 'submit'
}

export const Button = forwardRef<HTMLButtonElement, ButtonProps>((
  {
    autoFocus,
    className,
    children,
    disabled,
    icon: Icon,
    iconRight: IconRight,
    intent = 'base',
    onClick,
    loading,
    size = 'm',
    text,
    title,
    type = 'button',
    ...props
  }: ButtonProps,
  ref
) => {
  const isOnlyIcon = Icon && !children && !text;

  const content: JSX.Element[] = [];

  if (Icon) {
    content.push(
      <Icon
        key="Icon"
        className={cx(
          styles.icon,
          {
            [styles.iconMargin]: !isOnlyIcon
          }
        )}
      />
    )
  }

  content.push(<React.Fragment key="content">{children || text}</React.Fragment>)

  if (IconRight && !isOnlyIcon) {
    content.push(
      <IconRight
        key="IconRight"
        className={cx(
          styles.icon,
          styles.iconRightMargin
        )}
      />
    )
  }

  if (loading && !disabled) {
    content.push(
      <div key="loading" className={styles.loadingWrap}>
        <IconSpinner />
      </div>
    );
  }

  return (
    <button
      {...props}
      autoFocus={autoFocus}
      className={cx(
        styles.button,
        {
          [styles.iconicL]: isOnlyIcon && size === 'l',
          [styles.iconicM]: isOnlyIcon && size === 'm',
          [styles.iconicS]: isOnlyIcon && size === 's',
          [styles.l]: !isOnlyIcon && size === 'l',
          [styles.m]: !isOnlyIcon && size === 'm',
          [styles.s]: !isOnlyIcon && size === 's',
          [styles.xs]: size === 'xs',
          [intentActiveStyles[intent]]: !loading && !disabled
        },
        intentStyles[intent],
        {
          [styles.loading]: loading && !disabled
        },
        className
      )}
      disabled={disabled}
      onClick={onClick}
      title={title}
      type={type}
      ref={ref}
    >
      {content}
    </button>
  );
});
