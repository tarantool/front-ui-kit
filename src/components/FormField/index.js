// @flow
import * as React from 'react';
import { cx } from '@emotion/css';
import { Text } from '../Text';
import { Tooltip } from '../Tooltip';
import { IconInfo } from '../Icon';
import { ControlsPanel } from '../ControlsPanel';
import { InputGroup } from '../InputGroup';
import { commonFormFieldStyles as commonStyles } from './commonStyles';

type FormFieldProps = {
  children?: React.Node,
  className?: string,
  columns?: 1 | 2 | 3,
  info?: string | React.Node,
  itemClassName?: string,
  label?: string,
  largeMargins?: bool,
  subTitle?: string | React.Node,
  topRightControls?: React.Node[],
  verticalSort?: boolean
};

export const FormField = ({
  children,
  topRightControls,
  itemClassName,
  className,
  subTitle,
  columns,
  info,
  label,
  largeMargins,
  verticalSort
}:
FormFieldProps) => (
  <div
    className={cx(
      commonStyles.wrap,
      { [commonStyles.wrapMargin]: largeMargins },
      className
    )}
  >
    <div
      className={cx(
        commonStyles.headingPane,
        { [commonStyles.headingPaneMargin]: largeMargins }
      )}
    >
      <div className={commonStyles.headingPaneLeft}>
        {label && (
          <Text className={commonStyles.label} variant='h4' tag='span'>{label}</Text>
        )}
        {!!info && (
          <Tooltip className={commonStyles.tooltip} content={info}>
            <IconInfo />
          </Tooltip>
        )}
        {!!subTitle && (
          <Text className={commonStyles.subTitle} variant='p' tag='span'>{subTitle}</Text>
        )}
      </div>
      {topRightControls && (
        <ControlsPanel className={commonStyles.topRightControls} controls={topRightControls} />
      )}
    </div>
    <InputGroup columns={columns} itemClassName={itemClassName} verticalSort={verticalSort}>{children}</InputGroup>
  </div>
);