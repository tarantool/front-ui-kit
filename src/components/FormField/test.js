import renderer from 'react-test-renderer';
import React from 'react';
import { Button } from '../Button';
import { FormField } from '../FormField';
import { RadioButton } from '../RadioButton';

describe('FormField', () => {
  it('renders correctly', () => {
    const handleChange = jest.fn();

    const vshard_groups = [
      { name: 'group 1' },
      { name: 'group 2' },
      { name: 'group 3' },
      { name: 'group 4' }
    ];

    const tree = renderer.create(
      <FormField
        label='VShard group'
        info='Extended text description'
        subTitle='Subtitle'
        topRightControls={[
          <Button text='Right' />,
          <Button text='Contorls' />,
          <Button text='Pane' />
        ]}
      >
        {vshard_groups && vshard_groups.map(({ name }) => (
          <RadioButton
            key={name}
            onChange={handleChange}
            name='vshard_group'
            value={name}
            checked={name === 'group 2'}
          >
            {name}
          </RadioButton>
        ))}
      </FormField>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
