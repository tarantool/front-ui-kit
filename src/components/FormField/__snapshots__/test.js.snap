// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField renders correctly 1`] = `
<div
  className="css-pr10xp"
>
  <div
    className="css-1fg5of0"
  >
    <div
      className="css-1blnrmq"
    >
      <span
        className="css-1vk0mui"
      >
        VShard group
      </span>
      <div
        className="css-1ke4fav"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onScroll={[Function]}
        placement="top"
      >
        <svg
          className="css-1f5lesb"
          viewBox="0 0 16 16"
        >
          <use
            xlinkHref="#info"
          />
        </svg>
      </div>
      <span
        className="css-xy38qa"
      >
        Subtitle
      </span>
    </div>
    <div
      className="css-wb214m"
    >
      <div
        className="css-13w52vd"
      >
        <button
          className="css-1mocbyz"
          type="button"
        >
          Right
        </button>
      </div>
      <div
        className="css-13w52vd"
      >
        <button
          className="css-1mocbyz"
          type="button"
        >
          Contorls
        </button>
      </div>
      <div
        className="css-13w52vd"
      >
        <button
          className="css-1mocbyz"
          type="button"
        >
          Pane
        </button>
      </div>
    </div>
  </div>
  <div
    className="css-17bfre1"
  >
    <div
      className="css-1m4nrnu"
    >
      <label
        className="css-1uoamx5"
      >
        <input
          checked={false}
          className="css-ve0h83"
          name="vshard_group"
          onChange={[MockFunction]}
          type="radio"
          value="group 1"
        />
        <div
          className="css-qqecie"
        >
          <svg
            className="css-6i3pkg"
            viewBox="0 0 16 16"
          >
            <use
              xlinkHref="#radio"
            />
          </svg>
        </div>
        <span
          className="css-1ybmbi2"
        >
          group 1
        </span>
      </label>
    </div>
    <div
      className="css-1m4nrnu"
    >
      <label
        className="css-1uoamx5"
      >
        <input
          checked={true}
          className="css-ve0h83"
          name="vshard_group"
          onChange={[MockFunction]}
          type="radio"
          value="group 2"
        />
        <div
          className="css-qqecie"
        >
          <svg
            className="css-6i3pkg"
            viewBox="0 0 16 16"
          >
            <use
              xlinkHref="#radio-checked"
            />
          </svg>
        </div>
        <span
          className="css-1ybmbi2"
        >
          group 2
        </span>
      </label>
    </div>
    <div
      className="css-1m4nrnu"
    >
      <label
        className="css-1uoamx5"
      >
        <input
          checked={false}
          className="css-ve0h83"
          name="vshard_group"
          onChange={[MockFunction]}
          type="radio"
          value="group 3"
        />
        <div
          className="css-qqecie"
        >
          <svg
            className="css-6i3pkg"
            viewBox="0 0 16 16"
          >
            <use
              xlinkHref="#radio"
            />
          </svg>
        </div>
        <span
          className="css-1ybmbi2"
        >
          group 3
        </span>
      </label>
    </div>
    <div
      className="css-1m4nrnu"
    >
      <label
        className="css-1uoamx5"
      >
        <input
          checked={false}
          className="css-ve0h83"
          name="vshard_group"
          onChange={[MockFunction]}
          type="radio"
          value="group 4"
        />
        <div
          className="css-qqecie"
        >
          <svg
            className="css-6i3pkg"
            viewBox="0 0 16 16"
          >
            <use
              xlinkHref="#radio"
            />
          </svg>
        </div>
        <span
          className="css-1ybmbi2"
        >
          group 4
        </span>
      </label>
    </div>
  </div>
</div>
`;
